src/ecs/systems/audio.rs

```diff
@@ -0,0 +1,122 @@
+use std::iter::Iterator;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::Duration;
+
+use cgmath::{Matrix4, Transform, Point3};
+use ecs::{Join, System, ReadStorage, WriteStorage, Entity};
+use ecs::components::{AudioEmitter, AudioListener};
+use ecs::components::Transform as TransformComponent;
+use rodio::{SpatialSink, Source, Sample};
+
+/// Syncs 3D transform data with the audio engine to provide 3D audio.
+pub struct AudioSystem {
+    /// Entity which contains the AudioListener component that should be played for the player.
+    pub listener: Entity,
+}
+
+// Wraps a source and signals to another thread when that source has ended.
+struct EndSignalSource<I: Source>
+where
+    <I as Iterator>::Item: Sample,
+{
+    input: I,
+    signal: Arc<AtomicBool>,
+}
+
+impl<I: Source> EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    pub fn new(input: I, signal: Arc<AtomicBool>) -> EndSignalSource<I> {
+        EndSignalSource { input, signal }
+    }
+}
+
+impl<I: Source> Iterator for EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    type Item = <I as Iterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.input.next();
+        if next.is_none() {
+            self.signal.store(true, Ordering::Relaxed);
+        }
+        next
+    }
+}
+
+impl<I: Source> Source for EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    fn current_frame_len(&self) -> Option<usize> {
+        self.input.current_frame_len()
+    }
+
+    fn channels(&self) -> u16 {
+        self.input.channels()
+    }
+
+    fn samples_rate(&self) -> u32 {
+        self.input.samples_rate()
+    }
+
+    fn total_duration(&self) -> Option<Duration> {
+        self.input.total_duration()
+    }
+}
+
+impl<'a> System<'a> for AudioSystem {
+    type SystemData = (ReadStorage<'a, TransformComponent>,
+     ReadStorage<'a, AudioListener>,
+     WriteStorage<'a, AudioEmitter>);
+
+    fn run(&mut self, (transform, listener, mut audio_emitter): Self::SystemData) {
```

Comment:
Looks amazing, especially because it uses `Transform` now! Have to look at this code later.